crates/ibc/src/core/context.rs

```diff
@@ -120,12 +120,12 @@ impl std::error::Error for ContextError {
     }
 }
 
-pub trait Router {
+pub trait Router<'m> {
     /// Returns a reference to a `Module` registered against the specified `ModuleId`
-    fn get_route(&self, module_id: &ModuleId) -> Option<&dyn Module>;
+    fn get_route(&self, module_id: &ModuleId) -> Option<&(dyn Module + 'm)>;
 
     /// Returns a mutable reference to a `Module` registered against the specified `ModuleId`
-    fn get_route_mut(&mut self, module_id: &ModuleId) -> Option<&mut dyn Module>;
+    fn get_route_mut(&mut self, module_id: &ModuleId) -> Option<&mut (dyn Module + 'm)>;
```

Comment:
Boxing a reference seems odd; I suspect that this doesn't actually solve the problem if you want to allow to return `Module`s with a non-'static lifetime from `get_route`. In other words, it is not sufficient to remove the `'static` bound from `Module`, you also have to let them be return-able from the `Router`; and I'd be surprised that there'd be a way to do that without tracking the lifetimes?

I'll look for a counter-example to better show what I mean

Reply:
Actually, the best solution is simpler than my implementation and yours: it's to simply **not change anything**! Yes, very silly of me... :

```rust
pub trait Router {
    fn get_route(&self, module_id: &ModuleId) -> Option<&dyn Module>;
}
```

which, without any lifetimes elided, is equivalent to:

```rust
pub trait Router {
    fn get_route<'s, 'a>(&'s self, module_id: &'a ModuleId) -> Option<&'s (dyn Module + 's)>;
}
```

The crucial element that makes it work is the `+ 's` in `dyn Module + 's`; that is, by default, `dyn Module`s have the proper lifetime bound (especially *not* `'static`)!

## Why my solution worked
My solution worked because implicitly, `'m < 's` (`'s` outlives `'m`). So implementations that return a `&'s (dyn Module + 's)` satisfy the requirement (probably all real-world implementations).

But really the `'m` is unneeded and can simply be removed.

## Why your solution worked
Your solution, without lifetime elision, expands to:

```rust
pub trait Router {
    fn get_route<'s, 'a>(&'s self, module_id: &'a ModuleId) -> Option<&'s Box<dyn Module + 's>>;
}
```

The `dyn Module`s have a `'s` lifetime bound, which is what we want; the `Box` is not necessary and can be removed.

## How to fix this 
I think the PR should have been as simple as removing the `AsAnyMut` bound on `Module` (and fixing the tests), and not touching the `Router` at all... Yes, very silly. I'll try that on Monday!

Reply:
Thanks for looking into it. Glad we can keep things simple 👌